rolling-shutter/keyperimpl/gnosis/newslot.go

```diff
@@ -145,57 +161,56 @@ func (kpr *Keyper) isProposerRegistered(ctx context.Context, slot uint64, block
+	var txPointerAge int64
+	var txPointerOutdated bool
 	txPointerDB, err := gnosisKeyperDB.GetTxPointer(ctx, eon)
 	if err == pgx.ErrNoRows {
-		// The tx pointer is expected to be missing from the db if the eon has just started. In
-		// this case, we should initialize it to zero with an age of 1, ie decrypt starting with
-		// the first transaction.
-		// The tx pointer may also be missing if the keyper has been started late and no decryption
-		// key has been generated or received yet (receiving the keys message would update the
-		// pointer). In this case, the true age is unknown, as we only know the start block but
-		// not the start slot of the eon. However, we can ignore this edge case as it will be
-		// resolved automatically when the first keys message is received. If key generation
-		// continues to fail, eventually our tx pointer age will exceed the maximum value and we
-		// will start participating in the recovery process, albeit a bit late.
-		err := gnosisKeyperDB.SetTxPointer(ctx, gnosisdatabase.SetTxPointerParams{
-			Eon:   eon,
-			Slot:  slot,
+		log.Info().Int64("eon", eon).Msg("initializing tx pointer")
+		// If there is no tx pointer in the db, we initialize it to 0. This is the intended case
+		// for newly started eons. We might also end up doing that if the keyper has been started
+		// late for an eon. In this case, we would ideally set it to infinity (as we would do on a
+		// restart). 0 is ok though too.
+		err = gnosisKeyperDB.SetTxPointer(ctx, gnosisdatabase.SetTxPointerParams{
+			Eon: eon,
+			Age: sql.NullInt64{
+				Int64: 0,
+				Valid: true,
+			},
 			Value: 0,
 		})
 		if err != nil {
-			return 0, errors.Wrap(err, "failed to initialize tx pointer")
+			return 0, errors.Wrapf(err, "failed to initialize tx pointer for eon %d", eon)
 		}
 		txPointer = 0
-		txPointerAge = 1
+		txPointerAge = 0
+		txPointerOutdated = false
 	} else if err != nil {
 		return 0, errors.Wrap(err, "failed to query tx pointer from db")
 	} else {
 		txPointer = txPointerDB.Value
-		txPointerAge = slot - txPointerDB.Slot
-	}
-	if txPointerAge == 0 {
-		// A pointer of age 0 means we already received the pointer from a DecryptionKeys message
-		// even though we haven't sent our shares yet. In that case, sending our shares is
-		// unnecessary.
-		return 0, errZeroTxPointerAge
+		txPointerAge = txPointerDB.Age.Int64
+		if txPointerDB.Age.Valid {
+			txPointerOutdated = txPointerAge > maxTxPointerAge
+		} else {
+			txPointerOutdated = true
+		}
 	}
+
 	// If the tx pointer is outdated, the system has failed to generate decryption keys (or at
 	// least we haven't received them). This either means not enough keypers are online or they
 	// don't agree on the current value of the tx pointer. In order to recover, we choose the
 	// current length of the transaction queue as the new tx pointer, as this is a value
 	// everyone can agree on.
-	isOutdated := txPointerAge > maxTxPointerAge
-	if isOutdated {
+	if txPointerOutdated {
```

Comment:
Just so I understand correctly: if `txPointerOutdated` we do not update the value in the `db`, right? When will it be written again, if ever?

Reply:
I can answer myself: https://github.com/shutter-network/rolling-shutter/pull/457/files?diff=split&w=0#diff-ccc4cbc95d87e9313b03d6fe7faa8ed07de07ae087cd6969446b781b1915bb2fR269